### PR TITLE
[BUILD] Improve how to handle yield() in ARM.

### LINKED ISSUE #3128: Fails to build on ARMEL
**Describe your environment** 
Debian unstable on armel (32-bit little-endian ARM architecture using the Embedded ABI - v5te instruction set)

**Steps to reproduce**
Build opentelemetry-cpp on armel.

**What is the expected behavior?**
Successful build.

**What is the actual behavior?**

The build fails because of:

```
[  8%] Building CXX object api/test/trace/CMakeFiles/api_provider_test.dir/provider_test.cc.o
cd /home/santiago/opentelemetry-cpp-1.17.0/obj-arm-linux-gnueabi/api/test/trace && /usr/bin/c++ -DHAVE_ABSEIL -DOPENTELEMETRY_ABI_VERSION_NO=1 -I/home/santiago/opentelemetry-cpp-1.17.0/obj-arm-linux-gnueabi/generated/third_party/opentelemetry-proto -I/home/santiago/opentelemetry-cpp-1.17.0/api/include -g -O2 -ffile-prefix-map=/home/santiago/opentelemetry-cpp-1.17.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2 -Wno-error=deprecated-declarations -DGTEST_HAS_PTHREAD=1 -MD -MT api/test/trace/CMakeFiles/api_provider_test.dir/provider_test.cc.o -MF CMakeFiles/api_provider_test.dir/provider_test.cc.o.d -o CMakeFiles/api_provider_test.dir/provider_test.cc.o -c /home/santiago/opentelemetry-cpp-1.17.0/api/test/trace/provider_test.cc 
/tmp/cc6SyTzU.s: Assembler messages:
/tmp/cc6SyTzU.s:18172: Error: selected processor does not support `yield' in ARM mode
/tmp/cc6SyTzU.s:22037: Error: selected processor does not support `yield' in ARM mode
make[3]: *** [api/test/trace/CMakeFiles/api_provider_test.dir/build.make:79: api/test/trace/CMakeFiles/api_provider_test.dir/provider_test.cc.o] Error 1
```

You can see a build full log (with parallel jobs) at: https://buildd.debian.org/status/fetch.php?pkg=opentelemetry-cpp&arch=armel&ver=1.17.0-1&stamp=1730821961&raw=0

**Additional context**
N/A

I will propose a fix for this.

### PR DESCRIPTION
**Pull request by santiagorr**
## Changes

The yield instruction was introduced in ARM processors more recent than armel. So building in armel ends up in "Error: selected processor does not support `yield' in ARM mode". Also, the __yield() intrinsic instruction is not understood for armel by g++. So let's do nothing for armel.

This armel build CI job has the patch applied: https://salsa.debian.org/science-team/opentelemetry-cpp/-/jobs/6543996

Ref: https://developer.arm.com/documentation/ddi0406/b/Application-Level-Architecture/Application-Level-Programmers--Model/Exceptions--debug-events-and-checks/The-Yield-instruction?lang=en

Fixes #3128 

P.S. Example of other project handling similar instructions in different architectures: https://github.com/geidav/spinlocks-bench/blob/master/os.hpp#L31

### CODE REVIEW

**Review by lalitb — APPROVED** [review]
LGTM if it is working fine with this patch for ARMel in the CI as mentioned in the description :)

File: api/include/opentelemetry/common/spin_lock_mutex.h [hunk 1]
```diff
@@ -68,8 +68,10 @@ class SpinLockMutex
 #  else
     __builtin_ia32_pause();
 #  endif
-#elif defined(__arm__)
-    __asm__ volatile("yield" ::: "memory");
+#elif defined(__armel__) || defined(__ARMEL__)
+    asm volatile("nop" ::: "memory");
```
Comment (ThomsonTan) [inline api/include/opentelemetry/common/spin_lock_mutex.h:72]: Is there any difference between `asm volatile() and `__asm__ __volatile__()`?

Comment (santiagorr) [reply api/include/opentelemetry/common/spin_lock_mutex.h:72]: AFAIU, there is no difference between them, and it mostly depends on the projects conventions and style. Given that there are other occurrences of  `__asm__ volatile(...)`, and zero occurrences of  `asm volatile(...)` in other parts of the code, I might change it if you want.

**Review by marcalff — APPROVED** [review]
LGTM, based on the external armel build CI job provided.

Thanks for the fix.

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**